purchases/src/main/kotlin/com/revenuecat/purchases/SyncPendingTransactionsHelper.kt

```diff
@@ -0,0 +1,114 @@
+                            log(
+                                LogIntent.DEBUG,
+                                RestoreStrings.QUERYING_PURCHASE_WITH_HASH.format(purchase.type, hash),
+                            )
+                        }
+                        deviceCache.cleanPreviouslySentTokens(purchasesByHashedToken.keys)
+                        val transactionsToSync = deviceCache.getActivePurchasesNotInCache(purchasesByHashedToken)
+                        postTransactionsWithCompletion(
+                            transactionsToSync,
+                            allowSharingPlayStoreAccount,
+                            appUserID,
+                            onError,
+                            onSuccess,
+                        )
+                    },
+                    onError = { error ->
+                        log(LogIntent.GOOGLE_ERROR, error.toString())
+                        onError?.invoke(error)
+                    },
+                )
+            })
+        } else {
+            log(LogIntent.DEBUG, PurchaseStrings.BILLING_CLIENT_NOT_CONNECTED)
+            onError?.invoke(PurchasesError(PurchasesErrorCode.StoreProblemError, "Billing client disconnected"))
+        }
+    }
+
+    private fun postTransactionsWithCompletion(
+        transactionsToSync: List<StoreTransaction>,
+        allowSharingPlayStoreAccount: Boolean,
+        appUserID: String,
+        onError: ((PurchasesError) -> Unit)? = null,
+        onSuccess: ((CustomerInfo?) -> Unit)? = null,
+    ) {
+        if (transactionsToSync.isEmpty()) {
+            log(LogIntent.DEBUG, PurchaseStrings.NO_PENDING_PURCHASES_TO_SYNC)
+            onSuccess?.invoke(null)
+        } else {
+            val results: MutableList<Result<CustomerInfo, PurchasesError>> = mutableListOf()
+            postTransactionWithProductDetailsHelper.postTransactions(
+                transactionsToSync,
+                allowSharingPlayStoreAccount,
+                appUserID,
+                transactionPostSuccess = { _, customerInfo ->
+                    results.add(Result.Success(customerInfo))
+                    callCompletionFromResults(transactionsToSync, results, onError, onSuccess)
+                },
+                transactionPostError = { _, purchasesError ->
+                    results.add(Result.Error(purchasesError))
+                    callCompletionFromResults(transactionsToSync, results, onError, onSuccess)
+                },
+            )
+        }
+    }
+
+    private fun callCompletionFromResults(
+        transactionsToSync: List<StoreTransaction>,
+        results: List<Result<CustomerInfo, PurchasesError>>,
+        onError: ((PurchasesError) -> Unit)? = null,
+        onSuccess: ((CustomerInfo?) -> Unit)? = null,
+    ) {
+        if (transactionsToSync.size == results.size) {
+            results.forEachIndexed { index, result ->
+                if (result is Result.Error) {
+                    onError?.invoke(result.value)
```

Comment:
Note how I'm calling the error callback if any of the requests in the sync fails.